Source/Language/IReturns.cs

```diff
@@ -67,6 +68,26 @@ public partial interface IReturns<TMock, TResult> : IFluentInterface
 		IReturnsResult<TMock> Returns(TResult value);
 
 		/// <summary>
+		/// Specifies a function that will calculate the value to return from an open generic method.
+		/// </summary>
+		/// <param name="valueFunction">
+		/// The function that will calculate the return value.
+		/// IPublicCallContext contains the closed generic method as well as the arguments passed in.
+		/// </param>
+		/// <example group="returns">
+		/// Return a calculated value when the method is called:
+		/// <code>
+		/// mock.Setup(x => x.Execute&lt;It.IsAnyType>())
+		///     .Returns(context => returnValues[0]);
+		/// </code>
+		/// The lambda expression to retrieve the return value is lazy-executed, 
+		/// meaning that its value may change depending on the moment the method 
+		/// is executed and the value the <c>returnValues</c> array has at 
+		/// that moment.
+		/// </example>
+		IReturnsResult<TMock> Returns(Func<IPublicCallContext, object> valueFunction);
```

Comment:
I did contemplate about adding this when I initially created it. Unfortunately for Open Generic methods where the return type is open generic, will require a `Func` that returns `object` no matter what, as at setup time the return type is not known.

But I definitely agree, that potentially allowing the existing workflow to not catch type mismatches at compile time is definitely a no-no. I will take a look at trying to extract the open generic interface so that the `Func<object>` return type is only available when in "return type open-generic" mode.

Worst case scenario may need to introduce (which I was contemplating initially) one of the following:

```
moq.SetupGeneric(...) 
// Then can provide different branch of IReturns, which includes a Return<Func<...,object>>()
```

or

```
moq.Setup(...).ReturnsGeneric()
// Overload that can accept a Func<...,object>
```

But I will first take another look at trying to infer `Func<...,It.AnyType>`, at which point I might be able to provide suitable overloads without breaking existing functionality.